### ISSUE TITLE
Fix the roll-off calculation

### ISSUE THREAD

**Issue by bmcfee** [bmcfee/resampy#4]
We're currently getting a bit of aliasing at the high end of the frequency range when down-sampling.  (Up-sampling appears to be clean.)  This is most likely due to either a bad specification of the roll-off frequency, or something slightly off in the sinc window computation.

**Comment by bmcfee** [comment]
[paging @dpwe ]

Here's the results of my re-implementation of SRC's sine sweep test.

![image](https://cloud.githubusercontent.com/assets/1190540/14397118/a5a67956-fda9-11e5-9b2a-dac2d0dfe771.png)

I'm pretty sure the interpolation code is correct, and everything comes down to the window calculation; the above was set to match the kaiser window parameters used in `sinc_best` mode for src.  Near as I can tell, all the differences here come down to setting the roll-off in the low-pass filter, which I'm almost certainly doing wrong.  [Relevant code is here](https://github.com/bmcfee/resample/blob/master/resample/core.py#L11-L26)

Any thoughts?

**Comment by dpwe** [comment]
Yes I have thoughts.  You have components (at -100 dB) at like 7500 Hz and
15000 Hz, similar to the linear interpolator, just they are lower energy.

Presumably if you had limited the colorbar to 100 dB you would not see
these components and you'd look OK.

I assume that for some colorbar ranges, distortion products become visible
for SRC and SciPy too.  Not sure what "ideal" is in this context - you
calculate the unlimited sinc over the full duration of the waveform?

It looks like your LPF has a stop band rejection of around -100 dB, whereas
the SRC one is 20-40 dB better.  I haven't read the JOS reference closely,
so I don't know what his "single analog LPF" is designed to be, but in
general the way you find an optimal N-point FIR filter to achieve X dB of
ripple in the passband and Y dB of rejection in the stop band, and with a
certain width of transition band, is with Remez/Parks-McClelland/iterative
design.  The sample points are then not the result of any closed-form
function, but simply some point in filter design space.

I think you mentioned a big table of parameters somewhere, that's likely
where they came from.   Then it's just a question of playing with the
filter length and transition band width until you get down to the stop band
rejection and passband ripple you can tolerate.  Although you want them to
be zero, they never really are.  These kind of LPFs are non-monotonic, even
in the pass band, but that's how you minimize peak passband error.  You can
make the passband ripples very, very shallow at the cost of
length/transition band/stop band rejection.

e.g. 129 point LPF with transition band from 0.45 to 0.55 of nyquist, equal
weighting of pass and stop bands, achieves better than 100 dB rejection in
stop band:

lpf = scipy.signal.remez(129, [0, 0.45, 0.55, 1], [1, 0], weight=[1, 1],
Hz=2)
W, H = scipy.signal.freqz(lpf)
plot(W/2/np.pi, 20*np.log10(np.abs(H)))

.. with passband ripple about +/- 0.000005:
plot(W/2/np.pi, (np.abs(H)))
ylim([0.99999, 1.00001])

using weight=[10, 1] makes the passband ripple ~ 3x smaller, and the stop
band rejection about 10 dB less.

  DAn.

**Comment by bmcfee** [comment]
> I assume that for some colorbar ranges, distortion products become visible
> for SRC and SciPy too.  Not sure what "ideal" is in this context - you
> calculate the unlimited sinc over the full duration of the waveform?

Sorry, I should have clarified the test.  "Ideal" is the sine sweep generated natively at the target sampling rate.  The down-sweep at the end is due to aliasing, since the source signal sweeps beyond nyquist at the target sampling rate.

The colorbars are all on the same scale (180dB) to match the SRC test page.

> It looks like your LPF has a stop band rejection of around -100 dB, whereas
> the SRC one is 20-40 dB better.

Yup.  Side note: I thought I was matching SRC's window parameters, but I think this isn't exactly right.  SRC has pre-computed filters stored in header files with comments at the top describing the parameters, but these don't seem to line up to the filter definitions (ie they have different lengths).  I suspect the filters were updated but the comments are stale.  At this point, I'm not really sure where SRC's filters come from, and I'm not particularly interested in matching them exactly, but I would like to reduce the aliasing artifacts in what we have here.

I also looked into [resample](https://github.com/S25RTTR-Aux/s25rttr/blob/master/s-c/resample-1.8.1/src/filterkit.c#L143-L148) but the filter construction code doesn't match the paper; it uses `2 * f_c * sin(f_c * t)/t` rather than `2 * f_c * sinc(f_c * t)`, so there's a missing factor of `f_c` in the divisor.  Otherwise, my window construction should be equivalent (if you supply a kaiser window instead of the default).




> but in
> general the way you find an optimal N-point FIR filter to achieve X dB of
> ripple in the passband and Y dB of rejection in the stop band, and with a
> certain width of transition band, is with Remez/Parks-McClelland/iterative
> design

... intriguing.  If I understand what's going on in the rest of your comment, it shouldn't be hard to adapt this method to fit into the rest of fractional index/linear interpolation scheme.  Do you know if the result of `scipy.signal.remez` is guaranteed to be symmetric (or easily relatable to its time reverse)?

**Comment by dpwe** [comment]
Yes, the result of remez is symmetric by construction; the optimization is
in the space of odd-length symmetric filters.

  DAn.

**Comment by bmcfee** [comment]
If we go this route, it seems like it would make sense to do as src does, and just have a few pre-computed filters (low/med/high quality).

I tried to use remez to construct filters similar to the current windowed sinc implementation, but it fails to converge once the filters get long.  Maybe I'm doing something wrong here.

**Comment by dpwe** [comment]
It's hard to make the filters converge if they're really long (and if the
passband is really narrow).  What does SRC use?  I'm imagining 32 sinc
lobes and 32 samples per lobe, so maybe a 2049 point filter?

Yeah, I'm getting the same thing.  Can't get it to converge beyond around
257 points.

Hmm, it depends a lot.  I guess maybe try playing with the transition
band.  This one is working for me:

lpf = scipy.signal.remez(1025, [0, 1/64.0, 2/64.0, 1], [1, 0], weight=[1,
1], Hz=2, maxiter=1000)

That gives 130 dB rejection.  But a 1 octave transition band is a bit more
than you can tolerate, though.

lpf = scipy.signal.remez(2049, [0, 3/128.0, 4/128.0, 1], [1, 0], weight=[1,
1], Hz=2, maxiter=1000) also converges.

lpf = scipy.signal.remez(4097, [0, 7/256.0, 8/256.0, 1], [1, 0], weight=[1,
1], Hz=2, maxiter=1000)  converged eventually.

  DAn.

**Comment by bmcfee** [comment]
> What does SRC use?

I reverse engineered the filters, and I can get a kaiser+sinc to match src's filters down to +- 1e-4 or so.  So I don't think they're doing anything fancy.  The filters are much longer than what you're describing: high quality has half-length of 340238, fastest has 2463.

Lengths here are calculated by `2 * num_zeros * L + 1`, where num_zeros is the number of zero-crossings in the sinc, and L is an over-sampling parameter that determines the accuracy of the linear interpolation.

**Comment by bmcfee** [comment]
Actually, JOS talks about Remez [here](https://ccrma.stanford.edu/~jos/resample/Implementation.html) at the bottom.

**Comment by dpwe** [comment]
Yeah, well, we observed that [Remez nonconvergence for filters of 1000+
points].

Remez/Parks-McClelland comes from a time when computers ran at human
speeds; if it didn't converge in 5 or 6 iterations, it wasn't worth much,
and a thousand-point FIR filter had no practical applications.

Now things are different.  You don't need such a highly-efficient (but
temperamental) solution.  You can probably throw the whole thing into CVX
or whatever to find a better solution.  I don't know quite how to express
the minimax constraint (minimizing the peak gain in the stop band), but the
general principle of high-dimensional gradient descent is clearly
applicable (and really all anyone does these days).  I bet it's like 4
lines in TensorFlow.

  DAn.

**Comment by bmcfee** [comment]
Intriguing... but probably way more work than I'm looking to invest at the moment.

As I see it, there are two big questions that need addressing:

1. Is the current implementation "correct" in the sense that it computes a reasonable approximation of what the paper describes?
2. What's a good API for the resampler?  If we do end up going the remez route (or something like it), then filters will need to be pre-computed and indexed discretely, rather than parameterized explicitly.  I'd like to figure out how to handle this sanely before committing to any particular implementation, just in case we decide to switch things up later on.

**Comment by dpwe** [comment]
>1.  Is the current implementation "correct" in the sense that it computes a
>     reasonable approximation of what the paper describes?

Yeah, I'd like to understand that too.  I think it has to be "correct"
if it has lower distortion energy than
plain linear interpolation, but your last plots were not quite
satisfying enough.

Remember that super-hi-fi audio is 24 bit = 144 dB SNR; in practice,
distortion more than 100 dB below the signal isn't worth worrying
about.  What happens if you plot those curves with 100 dB colorbar?

And it would be nice to understand the magic SRC window.  It's
possible he started with a kaiser+sinc, then did a little bit of
gradient descent; I'm not sure how, but that's probably not that hard
- look at the peaks in the FFT of the whole window, and explicitly
subtract out little bits of peak energy.

> 2. What's a good API for the resampler? If we do end up going the remez route
>     (or something like it), then filters will need to be pre-computed and
>     indexed discretely, rather than parameterized explicitly. I'd like to figure
>     out how to handle this sanely before committing to any particular
>     implementation, just in case we decide to switch things up later on.

It seems like having a few precomputed optimized settings, plus a way
to back off to parametric settings if people want to do something
explicit, is a fair solution.

   DAn.

**Comment by bmcfee** [comment]
> Remember that super-hi-fi audio is 24 bit = 144 dB SNR; in practice,
> distortion more than 100 dB below the signal isn't worth worrying
> about.  What happens if you plot those curves with 100 dB colorbar?

Everything looks fine if you ignore <-100dB.  Perceptually, I've gone through a bunch of real examples at 44kHz->$random_rate and plotted the difference in energy after resampling compared to scipy (full fft resample), and it's generally spot on until you hit the transition band.  Artifacts are inaudible, but they still stick in my craw.



> It seems like having a few precomputed optimized settings, plus a way
> to back off to parametric settings if people want to do something
> explicit, is a fair solution.

Agreed.  Probably this will end up with a comparable interface to src, along with some functions to generate new filters.  I'll ponder a bit on how exactly to pull this off.

Thanks for all the input!

**Comment by dpwe** [comment]
So is the problem that you can't match SRC's table-defined filter with
closed form?  If he included a pre-defined filter instead of
calculating it on the fly, that strongly suggests to me that it can't
be exactly generated from closed form, which suggests it was generated
through some kind of gradient-descent scheme that is too slow or bulky
(or manual) to include in the shipping code.   If you're trying to
reproduce his results without resorting to a stored table, I'd guess
that's futile.

It's expected that the result of an optimal gradient-descent scheme
would be very close to Kaiser (or other heuristic-but-near-optimal)
approach; that's why the heuristic approaches are acceptable.  But it
doesn't mean that the two windows were designed by similar approaches,
just that they both converged on (approximately) the same optimum.

But I'm sure you know this, so I'm not clear what you're looking for.
My gut is that distortion < 100 dB isn't possible with any kind of
material bug.  Also, I wouldn't expect any reasonable closed-form (but
heuristic, i.e. windowing) design approach to do better than this
without really huge overengineering.

  DAn.

**Comment by bmcfee** [comment]
> But I'm sure you know this, so I'm not clear what you're looking for.
> My gut is that distortion < 100 dB isn't possible with any kind of
> material bug.  Also, I wouldn't expect any reasonable closed-form (but
> heuristic, i.e. windowing) design approach to do better than this
> without really huge overengineering.

My gut tends to agree, but my brain is paranoid.

LIS above, I'm not so interested in matching SRC as in verifying that this implementation matches the spec in the paper.  I'm reasonably well convinced of the latter at this point, so any further optimizations (ie precomputed, better filter designs) can be left as future work.

**Comment by bmcfee** [comment]
As an alternative to remez, it might also be sensible to just grid-search (or spearmint, whatever) over the space of rolloff/beta parameters.  I didn't much fiddle with these in comparing to src, but it seems like a simpler route to tuning the implementation we already have.

**Comment by bmcfee** [comment]
Marking this as closed for now.

Merging #8 includes the API changes described above, and two pre-baked filters using kaiser-sinc.  The beta and roll-off parameters for the pre-baked filters were tuned by GP optimization through spearmint.

If we want to include better filter designs in the future, it's now easy to do so.
